Глейзер Роман

RomanGleyzer opened the pull request:
@masssha1308

RomanGleyzer committed:
Сравнение объектов

RomanGleyzer committed:
Рефакторинг тестов

masssha1308 commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -1,7 +1,9 @@
+
 public class ObjectComparison
 {
     [Test]
```
Неинформативное название и описание. Не ясно что проверяется. Давай поправим

masssha1308 commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -14,16 +16,14 @@ public void CheckCurrentTsar()
         var expectedTsar = new Person("Ivan IV The Terrible", 54, 170, 70,
```
Давай явно укажем что значат эти параметры через named arguments, чтобы не приходилось проваливаться в конструктор Person чтобы разобраться что есть что

masssha1308 commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -14,16 +16,14 @@ public void CheckCurrentTsar()
-        ClassicAssert.AreEqual(expectedTsar.Parent.Age, actualTsar.Parent.Age);
-        ClassicAssert.AreEqual(expectedTsar.Parent.Height, actualTsar.Parent.Height);
-        ClassicAssert.AreEqual(expectedTsar.Parent.Parent, actualTsar.Parent.Parent);
+        actualTsar.Should()
```
Давай поправим форматирование Fluent Assertions. Либо в одну строку, либо с правильными переносами для цепочки вызовов

masssha1308 commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -34,7 +34,10 @@ public void CheckCurrentTsar_WithCustomEquality()
+        // Какие недостатки у такого подхода?
+        // Недостатки подхода:
+        // 1) Требуется ручное обновление метода сравнения при добавлении новых свойств в класс Person,
+        // 2) Тест менее информативен при падении. Он просто укажет, что объекты не равны, не показывая какие именно свойства не совпали.
```
Могут ли быть проблемы с рекурсией?

RomanGleyzer replied:
> Могут ли быть проблемы с рекурсией?

Да, попробовал сделать так, чтобы у обоих сравниваемых объектов родитель указывал на самого себя. В итоге метод сравнения убежал в бесконечную рекурсию:

<img width="367" height="105" alt="image" src="https://github.com/user-attachments/assets/a5f1ef26-fa87-4d2c-a6ef-eb93ec7df2c9" />

<img width="365" height="74" alt="image" src="https://github.com/user-attachments/assets/10d285e7-9344-4d93-b78c-d115675622d3" />

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
-        ClassicAssert.IsFalse(new NumberValidator(17, 2, true).IsValidNumber("0.000"));
-        ClassicAssert.IsFalse(new NumberValidator(3, 2, true).IsValidNumber("-1.23"));
-        ClassicAssert.IsFalse(new NumberValidator(3, 2, true).IsValidNumber("a.sd"));
+    [TestCase(-1, 2, true, TestName = "Исключение, когда precision отрицательный")]
```
Много где testName избыточен, т.к. очевиден из переданных аргументов

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    [TestCase(5, -1, true, TestName = "Исключение, когда scale отрицательный")]
+    [TestCase(1, 2, true, TestName = "Исключение, когда scale больше precision")]
+    [TestCase(1, 1, true, TestName = "Исключение исключение, когда scale равен precision")]
+    public void Ctor_Should_Throw_On_Invalid_Arguments(int precision, int scale, bool onlyPositive)
```
Лучше разбить на отдельные тесты по категориям (InvalidPrecision, InvalidScale и т.д.) для улучшения читабельности

masssha1308 replied:
В нейминге лучше придерживаться шаблона Method_Scenario_ExpectedResult

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    public void Ctor_Should_Throw_On_Invalid_Arguments(int precision, int scale, bool onlyPositive)
+    {
+        var act = () => new NumberValidator(precision, scale, onlyPositive);
+        act.Should().Throw<ArgumentException>();
```
Можно еще текст ошибки уточнить

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    }
+
+    [TestCase(1, 0, true, TestName = "Не выбрасывает исключение на корректных элемент")]
+    public void Ctor_Should_Not_Throw_On_Valid_Arguments(int precision, int scale, bool onlyPositive)
```
можно переписать без testCase, т.к. он один

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+
+    [TestCase("", 17, 2, true, TestName = "Возвращает false, когда значение пустое")]
+    [TestCase(null, 17, 2, true, TestName = "Возвращает false, когда значение null")]
+    public void Should_Return_False_On_NullOrEmpty(string value, int precision, int scale, bool onlyPositive)
```
Лучше указывать имя метода в нейминге теста, т.к. в классе может быть несколько методов

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    [TestCase("4_4", 11, 3, true, TestName = "Возвращает false, когда число содержит символ подчёркивания")]
+    [TestCase("abcd_4", 10, 5, true, TestName = "Возвращает false, когда строка содержит буквы и подчёркивание")]
+    [TestCase("--1", 12, 2, true, TestName = "Возвращает false, когда число содержит двойной минус")]
+    public void Should_Return_False_On_InvalidFormat(string value, int precision, int scale, bool onlyPositive)
```
Тоже лучше разбить по категориям

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    public void IsValid_Respects_Scale(string input, int precision, int scale,
+        bool onlyPositive, bool expected)
+    {
+        var validator = new NumberValidator(10, scale, onlyPositive);
```
магическое число

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,107 @@
+    [TestCase(0, 2, true, TestName = "Исключение, когда precision равен нулю")]
+    [TestCase(5, -1, true, TestName = "Исключение, когда scale отрицательный")]
+    [TestCase(1, 2, true, TestName = "Исключение, когда scale больше precision")]
+    [TestCase(1, 1, true, TestName = "Исключение исключение, когда scale равен precision")]
```
Кое-где в TestName's есть очепятки, проверь, пожалуйста

RomanGleyzer committed:
Исправления в сравнении царей

ObjectComparison стал TsarComparison, добавил named arguments, поправил форматирование Fluent Assertions, выяснил, что могу быть проблемы с рекурсией

RomanGleyzer committed:
Исправление в рефакторинге тестов

Убрал избыточный TestName, поправил опечатки, стал придерживаться нейминга Method_Scenario_ExpectedResult, добавил уточнения текстов ошибок, разбил тесты по категориям и избавился от магического числа

masssha1308 commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -6,102 +6,204 @@ namespace HomeExercise.Tasks.NumberValidator;
+    #region Некорректная дробность
+    
+    [TestCase(5, -1, true)]
+    public void Ctor_InvalidScale_Negative_ThrowsArgumentException(int precision,  int scale, bool onlyPositive)
```
Если TestCase один, то лучше не писать через TestCase, а написать просто тест

masssha1308 commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -4,7 +4,7 @@
+public class TsarComparison
 {
     [Test]
     [Description("Проверка текущего царя")]
```
категорию теста после рефакторинга можно убрать
